appinventor/appengine/src/com/google/appinventor/client/editor/youngandroid/palette/YoungAndroidPalettePanel.java

```diff
@@ -61,6 +59,90 @@
  */
 public class YoungAndroidPalettePanel extends Composite implements SimplePalettePanel, ComponentDatabaseChangeListener {
 
+  private static ComponentCoverage componentCoverage = ComponentCoverage.getInstance();
+
+  /**
+   * The Filter interface is used by the palette panel to determine what components
+   * to show. By default, an identity filter is used (everything is shown). Other
+   * implementations may override the filter by calling {@link #setFilter(Filter, boolean)}.
+   *
+   * @author ewpatton@mit.edu (Evan W. Patton)
+   */
+  public interface Filter {
+    /**
+     * Tests whether the given component type should be shown in the palette.
+     * @param componentTypeName The component type to check.
+     * @return True if the component should be shown, otherwise false.
+     */
+    boolean shouldShowComponent(String componentTypeName);
+
+    /**
+     * Tests whether the extensions panel should be shown.
+     * @return True if extensions are allowed, otherwise false.
+     */
+    boolean shouldShowExtensions();
+  }
+
+  // Identity filter implementation
+  private static final Filter IDENTITY = new Filter() {
+    @Override
+    public boolean shouldShowComponent(String componentTypeName) {
+      return true;
+    }
+
+    @Override
+    public boolean shouldShowExtensions() {
+      return true;
+    }
+  };
+
+  //show ios components only implementation
+  private static final Filter SHOWIOSONLY = new Filter() {
+    @Override
+    public boolean shouldShowComponent(String componentTypeName) {
+      return componentCoverage.isIosCompatible(componentTypeName);
+    }
+
+    @Override
+    public boolean shouldShowExtensions() {
+      return true;
+    }
+  };
+
+  //show android components only implementation
+  private static final Filter SHOWANDROIDONLY  = new Filter() {
+    @Override
+    public boolean shouldShowComponent(String componentTypeName) {
+      return componentCoverage.isAndroidCompatible(componentTypeName);
+    }
+
+    @Override
+    public boolean shouldShowExtensions() {
+      return true;
+    }
+  };
+
+  //show both compatible only implementation
+  private static final Filter SHOWBOTHCOMPATIBLE = new Filter() {
+    @Override
+    public boolean shouldShowComponent(String componentTypeName) {
+      return (componentCoverage.isAndroidCompatible(componentTypeName) && componentCoverage.isIosCompatible(componentTypeName));
+    }
+
+    @Override
+    public boolean shouldShowExtensions() {
+      return true;
+    }
+  };
+
+  // The singleton instance of the palette panel
+  private static YoungAndroidPalettePanel INSTANCE;
+
+  private Filter filter = IDENTITY;
```

Comment:
We should name this appropriately because it's not clear what this filter is filtering for (versus `osFilter`, which is slightly more useful).

Reply:
Also, it's not clear to me that this is being used anywhere yet.